metadata_please/source_checkout.py

```diff
@@ -0,0 +1,237 @@
+"""
+Best-effort metadata extraction for "source checkouts" -- e.g. a local dir containing pyproject.toml.
+
+This is different from an (extracted) sdist, which *should* have a generated dist-info already.
+
+Prefers:
+- PEP 621 metadata (pyproject.toml)
+- Poetry metadata (pyproject.toml)
+- Setuptools static metadata (setup.cfg)
+
+Notably, does not read setup.py or attempt to emulate anything that can't be read staticly.
+"""
+import re
+from pathlib import Path
+
+try:
+    import tomllib as toml
+except ImportError:
+    import toml  # type: ignore[no-redef,unused-ignore]
+
+from configparser import NoOptionError, NoSectionError, RawConfigParser
+
+from packaging.utils import canonicalize_name
+
+from .types import BasicMetadata
+
+OPERATOR_RE = re.compile(r"([<>=~]+)(\d.*)")
+
+
+def combine_markers(*markers: str) -> str:
+    filtered_markers = [m for m in markers if m and m.strip()]
+    if len(filtered_markers) == 0:
+        return ""
+    elif len(filtered_markers) == 1:
+        return filtered_markers[0]
+    else:
+        return " and ".join(f"({m})" for m in filtered_markers)
+
+
+def merge_extra_marker(extra_name: str, value: str) -> str:
+    """Simulates what a dist-info requirement string would look like if also restricted to an extra."""
+    if ";" not in value:
+        return f'{value} ; extra == "{extra_name}"'
+    else:
+        a, _, b = value.partition(";")
+        a = a.strip()
+        b = b.strip()
+        c = f'extra == "{extra_name}"'
+        return f"{a} ; {combine_markers(b, c)}"
+
+
+def from_source_checkout(path: Path) -> bytes:
+    return (
+        from_pep621_checkout(path)
+        or from_poetry_checkout(path)
+        or from_setup_cfg_checkout(path)
+    )
+
+
+def from_pep621_checkout(path: Path) -> bytes:
+    """
+    Returns a metadata snippet (which is zero-length if this is none of this style).
+    """
+    try:
+        data = (path / "pyproject.toml").read_text()
+    except FileNotFoundError:
+        return b""
+    doc = toml.loads(data)
+
+    buf: list[str] = []
+    for dep in doc.get("project", {}).get("dependencies", ()):
+        buf.append(f"Requires-Dist: {dep}\n")
+    for k, v in doc.get("project", {}).get("optional-dependencies", {}).items():
+        extra_name = canonicalize_name(k)
+        buf.append(f"Provides-Extra: {extra_name}\n")
+        for i in v:
+            buf.append("Requires-Dist: " + merge_extra_marker(extra_name, i) + "\n")
+
+    return "".join(buf).encode("utf-8")
```

Comment:
It looks like tilde is allowed by standard python packaging. https://packaging.python.org/en/latest/specifications/version-specifiers/#examples

Do we need to convert any tildes in the pep621 into less than/greater than operators?

Reply:
No, but it probably means the poetry translation could `s/~/~=/` if they use the same rules.  I intend to revisit how the translation happens next big block of travel time I have.